coral-schema/src/main/java/com/linkedin/coral/schema/avro/MergeHiveSchemaWithAvro.java

```diff
@@ -0,0 +1,262 @@
+    Schema hivePrimitive = hivePrimitiveToAvro(primitive);
+    // if there was no matching Avro primitive, use the Hive primitive
+    Schema result = partner == null ? hivePrimitive : checkCompatibilityAndPromote(hivePrimitive, partner);
+    return shouldResultBeOptional ? SchemaUtilities.makeNullable(result) : result;
+  }
+
+  private Schema checkCompatibilityAndPromote(Schema schema, Schema partner) {
+    // TODO: Check if schema is compatible with partner
+    Schema extractedPartnerSchema = extractIfOption(partner);
+    switch (schema.getType()) {
+      case BYTES:
+        if (extractedPartnerSchema.getType().equals(Schema.Type.FIXED)) {
+          return partner;
+        }
+        return schema;
+      case STRING:
+        if (extractedPartnerSchema.getType().equals(Schema.Type.ENUM)) {
+          return partner;
+        }
+        return schema;
+      default:
+        return schema;
+    }
+  }
+
+  /**
+   * A {@link PartnerAccessor} which matches the requested field from a partner Avro struct by case insensitive
+   * field name match
+   */
+  private static class AvroPartnerAccessor implements PartnerAccessor<Schema, Schema.Field> {
+    private static final AvroPartnerAccessor INSTANCE = new AvroPartnerAccessor();
+
+    private static final Schema MAP_KEY = Schema.create(Schema.Type.STRING);
+
+    @Override
+    public Schema.Field fieldPartner(Schema partner, String fieldName) {
+      Schema schema = extractIfOption(partner);
+      return (schema.getType() == Schema.Type.RECORD) ? findCaseInsensitive(schema, fieldName) : null;
+    }
+
+    @Override
+    public Schema fieldType(Schema.Field partnerField) {
+      return partnerField.schema();
+    }
+
+    @Override
+    public Schema mapKeyPartner(Schema partner) {
+      Schema schema = extractIfOption(partner);
+      return (schema.getType() == Schema.Type.MAP) ? MAP_KEY : null;
+    }
+
+    @Override
+    public Schema mapValuePartner(Schema partner) {
+      Schema schema = extractIfOption(partner);
+      return (schema.getType() == Schema.Type.MAP) ? schema.getValueType() : null;
+    }
+
+    @Override
+    public Schema listElementPartner(Schema partner) {
+      Schema schema = extractIfOption(partner);
+      return (schema.getType() == Schema.Type.ARRAY) ? schema.getElementType() : null;
+    }
+
+    private Schema.Field findCaseInsensitive(Schema struct, String fieldName) {
+      Preconditions.checkArgument(struct.getType() == Schema.Type.RECORD);
+      // TODO: Optimize? This will be called for every struct field, we will run the for loop for every struct field
+      for (Schema.Field field : struct.getFields()) {
+        if (field.name().equalsIgnoreCase(fieldName)) {
+          return field;
+        }
+      }
+      return null;
+    }
+  }
+
+  private static Schema extractIfOption(Schema schema) {
+    if (isNullableType(schema)) {
+      return getOtherTypeFromNullableType(schema);
+    } else {
+      return schema;
+    }
+  }
```

Comment:
`SchemaUtilities` already has a method `extractIfOption`, we can import and use that